Mathlib/MeasureTheory/Integral/SobolevInequality.lean

```diff
@@ -0,0 +1,661 @@
+  · rw [← lmarginal_union μ f hf]
+    · rw [← insert_compl_insert hi]
+      rfl
+    rw [Finset.disjoint_singleton_left, not_mem_compl]
+    exact mem_insert_self i s
+  · -- the main nontrivial point is to check that an exponent `p` satisfying `0 ≤ p` and
+    -- `(#ι - 1) * p ≤ 1` is in the valid range for the inductive-step lemma
+    refine le_trans ?_ hp
+    gcongr
+    suffices (s.card : ℝ) + 1 ≤ #ι by linarith
+    rw [← card_add_card_compl s]
+    norm_cast
+    gcongr
+    have hi' : sᶜ.Nonempty := ⟨i, by rwa [Finset.mem_compl]⟩
+    rwa [← card_pos] at hi'
+
+end GridLines
+
+/-- The "grid-lines lemma" (not a standard name), stated with a general parameter `p` as the
+exponent.  Compare with `lintegral_prod_lintegral_pow_le`.
+
+For any finite dependent product `Π i : ι, A i` of sigma-finite measure spaces, for any
+nonnegative real number `p` such that `(#ι - 1) * p ≤ 1`, for any function `f` from `Π i, A i` into
+the extended nonnegative reals, we consider an associated "grid-lines quantity", the integral of an
+associated function from `Π i, A i` into the extended nonnegative reals.  The value of this function
+at `x : Π i, A i` is obtained by multiplying a certain power of `f` with the product, for each
+co-ordinate `i`, of a certain power of the integral of `f` along the "grid line" in the `i`
+direction through `x`.
+
+This lemma bounds the Lebesgue integral of the grid-lines quantity by a power of the Lebesgue
+integral of `f`. -/
+theorem lintegral_mul_prod_lintegral_pow_le {p : ℝ} (hp₀ : 0 ≤ p)
+    (hp : (#ι - 1 : ℝ) * p ≤ 1) {f : (∀ i : ι, A i) → ℝ≥0∞} (hf : Measurable f) :
+    ∫⁻ x, f x ^ (1 - (#ι - 1 : ℝ) * p) * ∏ i, (∫⁻ xᵢ, f (update x i xᵢ) ∂μ i) ^ p ∂.pi μ
+    ≤ (∫⁻ x, f x ∂.pi μ) ^ (1 + p) := by
+  cases isEmpty_or_nonempty (∀ i, A i)
+  · simp_rw [lintegral_of_isEmpty]; refine' zero_le _
+  inhabit ∀ i, A i
+  have H : (∅ : Finset ι) ≤ Finset.univ := Finset.empty_subset _
+  simpa [lmarginal_univ] using GridLines.T_lmarginal_antitone μ hp₀ hp hf H default
+
+/-- Special case of the grid-lines lemma `lintegral_mul_prod_lintegral_pow_le`, taking the extremal
+exponent `p = (#ι - 1)⁻¹`. -/
+theorem lintegral_prod_lintegral_pow_le [Nontrivial ι]
+    {p : ℝ} (hp : Real.IsConjExponent #ι p)
+    {f} (hf : Measurable f) :
+    ∫⁻ x, ∏ i, (∫⁻ xᵢ, f (update x i xᵢ) ∂μ i) ^ ((1 : ℝ) / (#ι - 1 : ℝ)) ∂.pi μ
+    ≤ (∫⁻ x, f x ∂.pi μ) ^ p := by
+  have h0 : (1:ℝ) < #ι := by norm_cast; exact Fintype.one_lt_card
+  have h1 : (0:ℝ) < #ι - 1 := by linarith
+  have h2 : 0 ≤ ((1 : ℝ) / (#ι - 1 : ℝ)) := by positivity
+  have h3 : (#ι - 1 : ℝ) * ((1 : ℝ) / (#ι - 1 : ℝ)) ≤ 1 := by field_simp
+  have h4 : p = 1 + 1 / (↑#ι - 1) := by field_simp; rw [mul_comm, hp.sub_one_mul_conj]
+  rw [h4]
+  convert lintegral_mul_prod_lintegral_pow_le μ h2 h3 hf using 2
+  field_simp
+
+/-! ## The Gagliardo-Nirenberg-Sobolev inequality -/
+
+variable {F : Type*} [NormedAddCommGroup F] [NormedSpace ℝ F] [CompleteSpace F]
```

Comment:
The completeness assumption here is probably not necessary, but removing this can definitely wait for another PR.

Reply:
Done in #14262